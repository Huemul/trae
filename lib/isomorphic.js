import 'isomorphic-fetch'

import Trae from './trae'

const trae = new Trae()

export default trae
