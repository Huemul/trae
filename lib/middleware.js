const identity  = response => response;
const rejection = err => Promise.reject(err);


export default class Middleware {
  constructor() {
    this._before  = [];
    this._after   = [];
    this._finally = [];
  }

  before(fn) {
    this._before.push(fn);
    return this._before.length - 1;
  }

  after(fulfill = identity, reject = rejection) {
    this._after.push({ fulfill, reject });
    return this._after.length - 1;
  }

  finally(fn) {
    this._finally.push(fn);
    return this._finally.length - 1;
  }

  resolveBefore(config) {
    return this._before.reduce((promise, task) => {
      promise = promise.then(task);
      return promise;
    }, Promise.resolve(config));
  }

  resolveAfter(err, response) {
    return this._after.reduce((promise, task) => {
      promise = promise.then(task.fulfill, task.reject);
      return promise;
    }, err ? Promise.reject(err) : Promise.resolve(response));
  }


  resolveFinally() {
    this._finally.forEach(task => task());
  }
}
