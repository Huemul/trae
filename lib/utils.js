import _merge from 'merge';

/**
 * Recursively merge objects
 *
 * @param {Object} objects to merge
 * @return {Object} the merged objects
 */
export function merge(...params)  {
  return _merge.recursive(true, ...params);
}

/**
 * Returns an object with the skipped properties
 *
 * @param {Object} obj the object to skip properties from
 * @param {[String]} keys keys of the properties to skip
 * @return {Object} the object with the properties skipped
 */
export function skip(obj, keys) {
  const skipped = {};
  Object.keys(obj).forEach((objKey) => {
    if (keys.indexOf(objKey) === -1) {
      skipped[objKey] = obj[objKey];
    }
  });
  return skipped;
}
