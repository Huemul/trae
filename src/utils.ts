type WithHeaders = { headers: unknown };

// TODO: do not use `any`
export const merge = (...objs: any[]) => {
  const headers = objs
    .filter((obj): obj is WithHeaders =>
      typeof obj === 'object' && obj ? Boolean(obj.headers) : false,
    )
    .map((obj) => obj.headers);

  return headers.length > 0
    ? Object.assign({}, ...objs, { headers: Object.assign({}, ...headers) })
    : Object.assign({}, ...objs);
};
