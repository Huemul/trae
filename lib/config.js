import { merge, skip } from './utils';


export default class Config {
  constructor(config = {}) {
    this._defaults = { headers: {} };
    this._config   = {};

    this.set(config);
  }

  mergeWithDefaults(...configParams) {
    const params = merge(...configParams);

    const config = merge(
      this._defaults, 
      this.skipNotUsedMethods(params.method),
      this._config[params.method], 
      params
    );

    if (
      typeof config.body === 'object' &&
      config.headers &&
      config.headers['Content-Type'] === 'application/json'
    ) {
      config.body = JSON.stringify(config.body);
    }
    return config;
  }
  
  skipNotUsedMethods(currentMethod){
    const notUsedMethods = [ 'delete', 'get', 'head', 'patch', 'post', 'put' ] 
      .filter(method => currentMethod !== method.toLowerCase());
    return skip(this._config, notUsedMethods);
  }

  
  set(config) {
    this._config = merge(this._config, config);
  }

  get() {
    return merge(this._defaults, this._config);
  }
}
