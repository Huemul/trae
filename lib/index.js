import { format as formatUrl } from './helpers/url-handler'
import { skip, merge } from './utils'
import Middleware from './middleware'
import Config from './config'
import responseHandler from './helpers/response-handler'
import { version } from '../package.json'

class Trae {
  constructor(config = {}) {
    this._middleware = new Middleware()
    this._config = new Config(skip(config, ['baseUrl']))

    this.baseUrl(config.baseUrl || '')
    this._setupMethodsWithBody()
    this._setupMethodsWithNoBody()
    this._setupMiddlewareMethods()

    this.version = version
  }

  create(config) {
    const instance = new this.constructor(merge(this.defaults(), config))
    const mapAfter = ({ fulfill, reject }) => instance.after(fulfill, reject)
    this._middleware._before.forEach(instance.before)
    this._middleware._after.forEach(mapAfter)
    this._middleware._finally.forEach(instance.finally)
    return instance
  }

  defaults(config) {
    if (typeof config === 'undefined') {
      const defaults = this._config.get()
      this.baseUrl() && (defaults.baseUrl = this.baseUrl())
      return defaults
    }
    this._config.set(skip(config, ['baseUrl']))
    config.baseUrl && this.baseUrl(config.baseUrl)
    return this._config.get()
  }

  baseUrl(baseUrl) {
    if (typeof baseUrl === 'undefined') {
      return this._baseUrl
    }
    this._baseUrl = baseUrl
    return this._baseUrl
  }

  request(config = {}) {
    config.method || (config.method = 'get')
    const mergedConfig = this._config.merge(config)
    const url = formatUrl(this._baseUrl, config.url, config.params)

    return this._fetch(url, mergedConfig)
  }

  _fetch(url, originalConfig) {
    const config = merge(originalConfig, {
      method: originalConfig.method.toUpperCase(),
    })
    const onceDone = (...args) => this._middleware.resolveFinally(...args)

    return this._middleware
      .resolveBefore(config)
      .then((c) => fetch(url, c))
      .then((res) => responseHandler(res, config))
      .then(
        (res) => this._middleware.resolveAfter(undefined, res),
        (err) => this._middleware.resolveAfter(err),
      )
      .then(
        (res) => Promise.resolve(onceDone(config, url)).then(() => res),
        (err) =>
          Promise.resolve(onceDone(config, url)).then(() =>
            Promise.reject(err),
          ),
      )
  }

  _setupMethodsWithNoBody() {
    const fetch = (path, config, method) => {
      const url = formatUrl(this._baseUrl, path, config.params)
      const mergedConfig = this._config.merge(config, { method, url })

      return this._fetch(url, mergedConfig)
    }

    this.get = (path, config = {}) => fetch(path, config, 'get')
    this.delete = (path, config = {}) => fetch(path, config, 'delete')
    this.head = (path, config = {}) => fetch(path, config, 'head')
  }

  _setupMethodsWithBody() {
    const defaultConfig = {
      headers: { 'Content-Type': 'application/json' },
    }

    this._config.set({ post: defaultConfig })
    this._config.set({ put: defaultConfig })
    this._config.set({ patch: defaultConfig })

    const fetch = (path, body, config, method) => {
      const url = formatUrl(this._baseUrl, path, config.params)
      const mergedConfig = this._config.merge(config, { body, method, url })

      return this._fetch(url, mergedConfig)
    }

    this.post = (path, body = {}, config = {}) =>
      fetch(path, body, config, 'post')
    this.put = (path, body = {}, config = {}) =>
      fetch(path, body, config, 'put')
    this.patch = (path, body = {}, config = {}) =>
      fetch(path, body, config, 'patch')
  }

  _setupMiddlewareMethods() {
    this.before = (...args) => this._middleware.before(...args)
    this.after = (...args) => this._middleware.after(...args)
    this.finally = (...args) => this._middleware.finally(...args)
  }
}

export default new Trae()
