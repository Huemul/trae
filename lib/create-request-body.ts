import { TraeSettings } from '../src/types';

function isJSON(config: TraeSettings) {
  const headers = config.headers;
  // @ts-ignore
  const key = Object.keys(headers).find(
    // @ts-ignore
    (header: any) => header.toLowerCase() === 'content-type',
  );
  // @ts-ignore
  return key && headers[key].toLowerCase() === 'application/json';
}

function createRequestBody(content: any, config: TraeSettings) {
  return isJSON(config) ? JSON.stringify(content) : content;
}

export default createRequestBody;
