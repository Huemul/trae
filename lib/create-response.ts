import { BodyType, TraeSettings } from '../src/types';

const isValidReader = (reader: string): reader is BodyType =>
  ['arrayBuffer', 'blob', 'formData', 'json', 'text', 'raw'].includes(reader);

function isFormData(body: unknown): body is FormData {
  return typeof FormData !== 'undefined' && body instanceof FormData;
}

interface TraeResponseErrorArgs {
  message: string;
  config: TraeSettings;
  response: Response;
}
class TraeResponseError extends Error {
  config: TraeSettings;
  response: Response;

  constructor({ message, config, response }: TraeResponseErrorArgs) {
    super(message);
    this.config = config;
    this.response = response;
  }
}

function deriveReader(response: Response, config: TraeSettings) {
  const { bodyType } = config;
  const { headers, body } = response;

  if (bodyType && isValidReader(bodyType)) {
    return bodyType;
  }

  const contentType = headers.get('Content-Type');

  if (contentType === 'application/json') {
    return 'json';
  }

  if (contentType === 'multipart/form-data' || isFormData(body)) {
    return 'formData';
  }

  if (body instanceof ArrayBuffer) {
    return 'arrayBuffer';
  }

  if (body instanceof Blob) {
    // TODO: Investigate edge cases
    //       https://stackoverflow.com/a/55271454/3377073
    return 'blob';
  }

  return 'text';
}

function parseResponse(response: Response, config: TraeSettings) {
  const reader = deriveReader(response, config);

  return reader === 'raw'
    ? Promise.resolve(response)
    : response[reader]().then((data: unknown) => ({
        status: response.status,
        statusText: response.statusText,
        data,
      }));
}

export default function createResponse(
  response: Response,
  config: TraeSettings,
) {
    if (response.ok) {
      return parseResponse(response, config);
    }

    const error = new TraeResponseError({
      message: response.statusText,
      config,
      response,
    });

    // TODO: Why isn't the parsed response part of the object we reject with?
    // @ts-ignore
    return parseResponse(response, config).then(() => Promise.reject(error));
  }
