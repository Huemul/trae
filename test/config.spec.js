/* global describe it expect */
import { merge } from '../lib/utils';
import Config    from '../lib/config';

const DEFAULT_HEADERS = {
  Accept: 'application/json, text/plain, */*',
  'Content-Type': 'application/json'
};

const DEFAULT_CONFIG = {
  xsrfCookieName: 'XSRF-TOKEN',
  xsrfHeaderName: 'X-XSRF-TOKEN'
};

const defaults = merge(DEFAULT_CONFIG, { headers: DEFAULT_HEADERS });

const configParams = {
  mode: 'no-cors',
  credentials: 'same-origin'
};

function getConfigWithParams() {
  return new Config(configParams);
}

describe('Config -> config', () => {
  it('initializes with the default fetch config', () => {
    const config = new Config();

    expect(config._config).toEqual({});
    expect(config._defaults).toEqual(defaults);
    expect(config.get()).toEqual(defaults);
  });

  it('initializes with the defaults merged with the provided config', () => {
    const config = getConfigWithParams();

    expect(config._config).toEqual(configParams);
    expect(config._defaults).toEqual(defaults);
    expect(config.get()).toEqual(merge(defaults, configParams));
  });

  describe('set', () => {

    it('merges the provided config with this._config', () => {
      const config = getConfigWithParams();
      config.set({
        mode: 'cors',
        headers: {
          'X-ACCESS-TOKE': 'aasdljhf2kjrasdf2l3jrhn2'
        }
      });

      expect(config._config).toMatchSnapshot();
    });

  });

  describe('get', () => {

    it('returns the merged _defaults and _config', () => {
      const config = getConfigWithParams();
      config.set({
        mode: 'cors',
        headers: {
          'X-ACCESS-TOKE': 'aasdljhf2kjrasdf2l3jrhn2'
        }
      });

      expect(config.get()).toMatchSnapshot();
    });

  });

  describe('mergeWithDefaults', () => {

    it('returns body stringified according to Content-Type', () => {
      const config = new Config();

      const actual = config.mergeWithDefaults({ body: { foo: 'bar' } });
      expect(actual).toMatchSnapshot();
    });

    it('returns the config merged with the params', () => {
      const config = new Config({ mode: 'no-cors' });

      const actual = config.mergeWithDefaults({ credentials: 'same-origin' });
      expect(actual).toMatchSnapshot();
    });

    it('returns the config merged with the params', () => {
      const config = new Config({ mode: 'no-cors' });

      const actual = config.mergeWithDefaults({ credentials: 'same-origin' });
      expect(actual).toMatchSnapshot();
    });

    it('merges all the params passed', () => {
      const config = new Config({ mode: 'no-cors' });

      const actual = config.mergeWithDefaults(
        { credentials: 'same-origin' },
        { headers: { 'Content-Type': 'text/plain' } }
      );
      expect(actual).toMatchSnapshot();
    });

    it('does not mutate _config or _defaults', () => {
      const config = new Config({ mode: 'no-cors' });

      const defaults = {
        headers: {
          Accept: 'application/json, text/plain, */*',
          'Content-Type': 'application/json'
        },
        xsrfCookieName: 'XSRF-TOKEN',
        xsrfHeaderName: 'X-XSRF-TOKEN'
      };

      expect(config._config).toEqual({ mode: 'no-cors' });
      expect(config._defaults).toEqual(defaults);

      const actual = config.mergeWithDefaults(
        { headers: { 'Content-Type': 'text/plain' } },
        { mode: 'cors' }
      );
      expect(actual).toMatchSnapshot();

      expect(config._config).toEqual({ mode: 'no-cors' });
      expect(config._defaults).toEqual(defaults);
    });

  });

});
