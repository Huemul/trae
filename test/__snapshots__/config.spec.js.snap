exports[`Config -> config get returns the merged _defaults and _config 1`] = `
Object {
  "credentials": "same-origin",
  "headers": Object {
    "X-ACCESS-TOKE": "aasdljhf2kjrasdf2l3jrhn2",
  },
  "mode": "cors",
}
`;

exports[`Config -> config merge does not mutate _config or _defaults 1`] = `
Object {
  "headers": Object {
    "Content-Type": "text/plain",
  },
  "mode": "cors",
}
`;

exports[`Config -> config merge merges all the params passed 1`] = `
Object {
  "credentials": "same-origin",
  "headers": Object {
    "Content-Type": "text/plain",
  },
  "mode": "no-cors",
}
`;

exports[`Config -> config merge returns body stringified according to Content-Type 1`] = `
Object {
  "body": Object {
    "foo": "bar",
  },
  "headers": Object {},
}
`;

exports[`Config -> config merge returns the config merged with the params 1`] = `
Object {
  "credentials": "same-origin",
  "headers": Object {},
  "mode": "no-cors",
}
`;

exports[`Config -> config merge returns the config merged with the params 2`] = `
Object {
  "credentials": "same-origin",
  "headers": Object {},
  "mode": "no-cors",
}
`;

exports[`Config -> config set merges the provided config with this._config 1`] = `
Object {
  "credentials": "same-origin",
  "headers": Object {
    "X-ACCESS-TOKEN": "aasdljhf2kjrasdf2l3jrhn2",
  },
  "mode": "cors",
}
`;
